File: app/src/main/java/com/jx/androiddemo/BaseMvpFragment.java
package com.jx.androiddemo;

import android.app.Activity;
import android.content.Context;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.Nullable;

import com.trello.rxlifecycle2.LifecycleTransformer;
import com.jx.androiddemo.di.component.DaggerFragmentComponent;
import com.jx.androiddemo.di.component.FragmentComponent;
import com.jx.androiddemo.di.module.FragmentModule;
import com.jx.androiddemo.event.NoticeEvent;
import com.jx.androiddemo.presenter.BasePresenter;
import com.jx.androiddemo.presenter.BaseRxPresenter;
import com.jx.androiddemo.tool.ToastUtil;

import javax.inject.Inject;

import butterknife.ButterKnife;
import butterknife.Unbinder;

/**
 * Created by WangTao
 */
public abstract class BaseMvpFragment<T extends BasePresenter> extends BaseAppFragment implements BaseView
{
    @Inject
    protected T mPresenter;

    protected View mView;

    protected Activity mActivity;

    protected Context mContext;

    private Unbinder mUnBinder;

    private boolean mInBack = true;

    @Override
    public void onAttach(Context context)
    {
        mActivity = (Activity) context;
        mContext = context;
        super.onAttach(context);
    }

    protected FragmentComponent getFragmentComponent()
    {
        return DaggerFragmentComponent.builder()
                .appComponent(BaseApplication.getAppComponent())
                .fragmentModule(getFragmentModule())
                .build();
    }

    protected FragmentModule getFragmentModule()
    {
        return new FragmentModule(this);
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState)
    {
        mView = inflater.inflate(getLayoutId(), null);
        initInject();
        return mView;
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState)
    {
        super.onViewCreated(view, savedInstanceState);
        if (savedInstanceState == null)
        {
            mUnBinder = ButterKnife.bind(this, view);
            if (mPresenter != null)
            {
                mPresenter.attachView(this);
            }
            initEventAndData();
        }
    }

    @Override
    public void onSaveInstanceState(Bundle outState)
    {
//        super.onSaveInstanceState(outState);
    }

    @Override
    public <T> LifecycleTransformer<T> bindToLife()
    {
        return this.bindToLifecycle();
    }

    @Override
    public void onDestroyView()
    {
        super.onDestroyView();
    }

    @Override
    public void onDestroy()
    {
        super.onDestroy();
        if (mPresenter != null)
        {
            mPresenter.detachView();
        }
        if (mUnBinder != null)
        {
            mUnBinder.unbind();
        }
        if (mPresenter != null && mPresenter instanceof BaseRxPresenter)
        {
            ((BaseRxPresenter) mPresenter).doDispose();
        }
    }

    protected abstract void initInject();

    protected abstract int getLayoutId();

    protected abstract void initEventAndData();

    @Override
    public void showMsg(String msg)
    {
        ToastUtil.showTextToast(msg);
    }

    @Override
    public void onEventMainThread(NoticeEvent event)
    {
    }

    @Override
    public void showProgressDialog(int tag, boolean canCancel)
    {
    }

    @Override
    public void hiddenProgressDialog()
    {
    }

    @Override
    public void onResume()
    {
        super.onResume();
        mInBack = false;
    }

    @Override
    public void onPause()
    {
        super.onPause();
        mInBack = true;
    }

    public boolean isInBack()
    {
        return mInBack;
    }
}

File: app/src/main/java/com/jx/androiddemo/BaseMvpFragment.kt
package com.jx.androiddemo

import android.app.Activity
import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.jx.androiddemo.event.NoticeEvent
import com.jx.androiddemo.presenter.BasePresenter
import com.jx.androiddemo.presenter.BaseRxPresenter
import me.yokeyword.fragmentation.SupportFragment
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode

abstract class BaseMvpFragment<T : BasePresenter> : SupportFragment(), BaseView {
    private var mPresenter: T? = null
    protected var mView: View? = null

    protected var mActivity: Activity? = null

    protected var mContext: Context? = null

    private var mInBack = true
    abstract fun initPresenter()
    override fun onAttach(context: Context) {
        super.onAttach(context)
        mActivity = context as Activity
        mContext = context
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        initPresenter()
        EventBus.getDefault().register(this)
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        mView = inflater.inflate(getLayoutId(), null)
        initInject()
        return mView
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    open fun onEventMainThread(event: NoticeEvent?) {
    }

    override fun onDestroy() {
        EventBus.getDefault().unregister(this)
        mPresenter?.detachView()
        if (mPresenter is BaseRxPresenter) {
            (mPresenter as BaseRxPresenter).doDispose()
        }
        super.onDestroy()
    }


    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        if (savedInstanceState == null) {
            initEventAndData()
            mPresenter?.attachView(this)
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
    }

    abstract fun initInject()

    abstract fun getLayoutId(): Int

    abstract fun initEventAndData()

    override fun onResume() {
        super.onResume()
        mInBack = false
    }

    override fun onPause() {
        super.onPause()
        mInBack = true
    }

    fun isInBack(): Boolean {
        return mInBack
    }

}